File: python/HuffmanTree.py
from Node import Node
class HuffmanTree:
    def __init__(self):
        self.root = None

    def createRoot(self, data, freq):
        self.root = Node(data, freq)

    def fromData(data, freq):
        newTree = HuffmanTree()
        newTree.createRoot(data, freq)
        return newTree

    def fromTrees(left, right):
        newTree = HuffmanTree()
        newTree.createRoot('\0', left.root.freq + right.root.freq)
        newTree.root.left = left.root
        newTree.root.right = right.root

        newTree.root.left.updateCode('0')
        newTree.root.right.updateCode('1')
        return newTree

    def __lt__(self, rhs):
        return self.root.freq < rhs.root.freq

    def characterCodes(self):
        charCodeMap = {}
        dfsStack = []
        dfsStack.append(self.root)
        while len(dfsStack):
            node = dfsStack.pop()
            if(node.isLeaf()):
                charCodeMap[node.data] = node.code
            else:
                dfsStack.append(node.left)
                dfsStack.append(node.right)
        
        return charCodeMap

File: cpp/HuffmanTree.cpp
#include "Node.cpp"
#include <string>
#include <map>
#include <stack>
using namespace std;
class HuffmanTree {
    public:
        Node* root;
        HuffmanTree(int data, int freq) {
            this->root = new Node(data, freq);
        }

        HuffmanTree(HuffmanTree left, HuffmanTree right) {
            this->root = new Node('\0', left.root->freq + right.root->freq);
            this->root->left = left.root;
            this->root->right = right.root;

            this->root->left->updateCode("0");
            this->root->right->updateCode("1");
        }
        
        map<char, string> characterCodes() {
            map<char, string> charCodeMap;
            stack<Node*> dfsStack;
            dfsStack.push(this->root);
            while(!dfsStack.empty()) {
                Node* node = dfsStack.top(); dfsStack.pop();
                if(node->isLeaf()) {
                    charCodeMap[node->data] = node->code;
                }else{
                    dfsStack.push(node->left);
                    dfsStack.push(node->right);
                }
            }
            return charCodeMap;
        }
};

class HuffmanTreeCompare {
    public:
        bool operator()(HuffmanTree a, HuffmanTree b){
            return a.root->freq > b.root->freq;
        }
};